geps/gep-1911.md

```diff
@@ -0,0 +1,238 @@
+- No easy mechanic to specify priority of protocols
+
+### Re-using *Route objects
+
+Route resources could be repurposed to also describe backend capabilities. For example we could have an HTTPRoute
+with a parentRef to an existing Kubernetes Service on a specific port. This would imply the that port's protocol
+is HTTP.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: httproute-example
+spec:
+  parentRefs:
+  - name: acme-lb
+    group: ""
+    kind: Service
+    port: 80
+```
+
+The advantages of this approach are:
+- Single resource can describe downstream and upstream protocol support
+- Protocol specific configuration needs to only be added to a single resource
+
+The disadvantages of this approach are:
+- Route objects are fairly complex (multiple parentRefs )- this would make it worse
+- Has same disadvantages as multiple protocol metaresources
+
+### Adding properties on Gateway Route Objects
+
+From [GEP-1282](geps/gep-1282.md#tldr):
+> some types of configuration requested by users are more about defining functionality that describes capabilities of the backend more than the route you take to get to the backend.
+
+Backend protocol is specifying capabilities. This configuration is less about routing.
+
+### Kubernetes Service - Expanding Protocol field
+
+The `protocol` field on a Kubernetes service is used to specify a L4 protocol over IP. This field isn't appropriate to describe protocols
+that operate at a higher level (eg. HTTP/GRPC etc.)
+
+### Kubernetes Service - Expanding AppProtocol field
+
+Since Kubernetes 1.20 the core/v1.Service object has a stable `appProtocol` field. It's purpose is to allow end-users to
+specify an application protocol (L7) for each service port.
+
+Originally the use of this field in the Gateway API was rejected in [GEP-1282](geps/gep-1282.md#non-goals):
+> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 
+
+Further discussion [here](https://github.com/kubernetes-sigs/gateway-api/discussions/1244#discussioncomment-3151197)
+
+Since then a Kubernetes enhancement proposal was created [KEP-3726](https://github.com/kubernetes/enhancements/issues/3726)
+to repurpose `appProtocol` to include a convention for protocols that are not IANA service names. This would involve
+prefixing protocol names with `kubernetes.io/`. At the moment there exists two defined protocols `kubernetes.io/h2c` and `kubernetes.io/grpc`.
+
+It is not possible to have different `appProtocol` entries for the same port. A related [KEP-1435](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb) mentions different `protocol` entries for the same port. Finally it's unclear how one would disable certain protocol behaviours (ie. HTTP/1 Upgrades).
```

Comment:
I know envoy has options to disable this in various places - https://www.envoyproxy.io/docs/envoy/v1.26.1/intro/arch_overview/http/upgrades.html

I don't have a use case really - just added this clause since `appProtocol` mechanic allows enabling but not disabling.

Reply:
I made this a non-goal for the GEP